File: app/src/main/java/org/oppia/app/databinding/TextViewBindingAdapters.java
package org.oppia.app.databinding;

import android.content.Context;
import android.content.res.Resources;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.annotation.PluralsRes;
import androidx.databinding.BindingAdapter;
import org.oppia.app.R;
import org.oppia.util.system.OppiaDateTimeFormatter;
import java.util.Locale;
import java.util.concurrent.TimeUnit;

/** Holds all custom binding adapters that bind to [TextView]. */
public final class TextViewBindingAdapters {

  /** Binds date text with relative time. */
  @BindingAdapter("profile:created")
  public static void setProfileDataText(@NonNull TextView textView, long timestamp) {
    OppiaDateTimeFormatter oppiaDateTimeFormatter = new OppiaDateTimeFormatter();
    String time = oppiaDateTimeFormatter.formatDateFromDateString(
        OppiaDateTimeFormatter.DD_MMM_YYYY,
        timestamp,
        Locale.getDefault()
    );
    textView.setText(textView.getContext().getString(
        R.string.profile_edit_created,
        time
    ));
  }

  @BindingAdapter("profile:lastVisited")
  public static void setProfileLastVisitedText(@NonNull TextView textView, long timestamp) {
    String profileLastUsed = textView.getContext().getString(R.string.profile_last_used);
    String timeAgoTimeStamp = getTimeAgo(
        timestamp,
        textView.getContext()
    );
    String profileLastVisited = textView.getContext().getString(
        R.string.profile_last_visited,
        profileLastUsed,
        timeAgoTimeStamp
    );
    textView.setText(profileLastVisited);
  }

  private static String getTimeAgo(long lastVisitedTimeStamp, Context context) {
    OppiaDateTimeFormatter oppiaDateTimeFormatter = new OppiaDateTimeFormatter();
    long timeStampMillis =
        oppiaDateTimeFormatter.checkAndConvertTimestampToMilliseconds(lastVisitedTimeStamp);
    long currentTimeMillis = oppiaDateTimeFormatter.currentDate().getTime();

    if (timeStampMillis > currentTimeMillis || timeStampMillis <= 0) {
      return "";
    }

    Resources res = context.getResources();
    long timeDifferenceMillis = currentTimeMillis - timeStampMillis;

    if (timeDifferenceMillis < (int) TimeUnit.MINUTES.toMillis(1)) {
      return context.getString(R.string.just_now);
    } else if (timeDifferenceMillis < TimeUnit.MINUTES.toMillis(50)) {
      return getPluralString(
          context,
          R.plurals.minutes,
          (int) TimeUnit.MILLISECONDS.toMinutes(timeDifferenceMillis)
      );
    } else if (timeDifferenceMillis < TimeUnit.DAYS.toMillis(1)) {
      return getPluralString(
          context,
          R.plurals.hours,
          (int) TimeUnit.MILLISECONDS.toHours(timeDifferenceMillis)
      );
    } else if (timeDifferenceMillis < TimeUnit.DAYS.toMillis(2)) {
      return context.getString(R.string.yesterday);
    }
    return getPluralString(
        context,
        R.plurals.days,
        (int) TimeUnit.MILLISECONDS.toDays(timeDifferenceMillis)
    );
  }

  private static String getPluralString(
      @NonNull Context context,
      @PluralsRes int pluralsResId,
      int count
  ) {
    Resources resources = context.getResources();
    return context.getString(
        R.string.time_ago,
        resources.getQuantityString(
            pluralsResId,
            count,
            count
        )
    );
  }
}

File: app/src/main/java/org/oppia/app/databinding/TextViewBindingAdapters.kt
package org.oppia.app.databinding

import android.content.Context
import android.widget.TextView
import androidx.databinding.BindingAdapter
import org.oppia.app.R
import org.oppia.util.system.OppiaDateTimeFormatter

/** Binds date text with relative time. */
@BindingAdapter("profile:created")
fun setProfileDataText(textView: TextView, timestamp: Long) {
  val oppiaDateTimeFormatter = OppiaDateTimeFormatter()
  val time = oppiaDateTimeFormatter.formatDateFromDateString(
    OppiaDateTimeFormatter.DD_MMM_YYYY,
    timestamp
  )
  textView.text = String.format(textView.context.getString(R.string.profile_edit_created, time))
}

@BindingAdapter("profile:lastVisited")
fun setProfileLastVisitedText(textView: TextView, timestamp: Long) {
  textView.text =
    String.format(
      textView.context.getString(R.string.profile_last_used) + " " + getTimeAgo(
        timestamp,
        textView.context
      )
    )
}

private const val SECOND_MILLIS = 1000
private const val MINUTE_MILLIS = 60 * SECOND_MILLIS
private const val HOUR_MILLIS = 60 * MINUTE_MILLIS
private const val DAY_MILLIS = 24 * HOUR_MILLIS

fun getTimeAgo(lastVisitedTimeStamp: Long, context: Context): String {

  val oppiaDateTimeFormatter = OppiaDateTimeFormatter()
  val timeStamp =
    oppiaDateTimeFormatter.checkAndConvertTimestampToMilliseconds(lastVisitedTimeStamp)
  val now = oppiaDateTimeFormatter.currentDate().time

  if (timeStamp > now || timeStamp <= 0) return ""

  val res = context.resources
  val timeDifference = now - timeStamp
  return when {
    timeDifference < MINUTE_MILLIS -> context.getString(R.string.just_now)
    timeDifference < 50 * MINUTE_MILLIS -> context.getString(
      R.string.time_ago,
      res.getQuantityString(
        R.plurals.minutes,
        timeDifference.toInt() / MINUTE_MILLIS,
        timeDifference / MINUTE_MILLIS
      )
    )
    timeDifference < 24 * HOUR_MILLIS -> context.getString(
      R.string.time_ago,
      res.getQuantityString(
        R.plurals.hours,
        timeDifference.toInt() / HOUR_MILLIS,
        timeDifference / HOUR_MILLIS
      )
    )
    timeDifference < 48 * HOUR_MILLIS -> context.getString(R.string.yesterday)
    else -> context.getString(
      R.string.time_ago,
      res.getQuantityString(
        R.plurals.days,
        timeDifference.toInt() / DAY_MILLIS,
        timeDifference / DAY_MILLIS
      )
    )
  }
}
